### DISCUSSION TITLE
Unable to send the traces to gateway url using auto instrumentation - giving Required SETTINGS preface not received when i use GRPC.

### DISCUSSION THREAD

**Discussion by ramanjaneyagupta** [open-telemetry/opentelemetry-java-instrumentation#8181] [Q&A]
Hi I am trying to send the traces to gateway( a set of collectors running behind the VIP), and when i am trying to send using GRPC it is giving below error - 
ERROR io.opentelemetry.exporter.internal.grpc.OkHttpGrpcExporter-Failedtoexportspans.Therequestcouldnotbe executed. Full error message: Required SETTINGS preface not received.
After a minute it is giving this :
ERROR io.opentelemetry.exporter.internal.grpc.OkhttpGrpcExporter-Failedtoexportspans.
The request could not be executed. Full error message: Broken pipe (Write failed)

Configuration: 
java -javaagent:/home/rpasumar/agent/opentelemetry-javaagent_1.24.jar\
-Dotel. traces.exporter=otlp
-Dotel.exporter.otlp.traces.endpoint=<gateway-url>:4317
-Dotel.exporter.otlp.traces.headers-"tIs.insecure=false"
-Dotel.service.name=petclinic
-Dserver.port=8081
-jar target/spring-petclinic-2.6.0-SNAPSHOT.jar

NOTE: if i have the collector(agent mode) installed on same vm and send the traces to that local collector, from local collector to gateway i didnt see any issue.but directly sending it to gateway it is giving above issue.

**Comment by trask** [comment]
hi @ramanjaneyagupta!

The error message you are seeing suggests that there is a problem with the GRPC communication between your application and the collectors running behind the VIP. Specifically, the error message "Required SETTINGS preface not received" indicates that the communication between the client and server is not being established properly.

Here are a few things you can try to resolve the issue:

* Verify that the endpoint address is correct and that the collectors are running and reachable from your application.
* Check that your firewall settings allow traffic on the port you are using for communication with the collectors.
* Increase the logging level to get more detailed information about the issue. You can add the following to your application's command line arguments to enable debug logging: `-Dio.opentelemetry.javaagent.slf4j.simpleLogger.defaultLogLevel=debug`. This will enable debug logging for the OpenTelemetry Java agent, which may provide more information about the issue.

**Reply by ramanjaneyagupta** [reply]
- Address is correct and collectors are running behind the VIP (if i try to send traces from local collector to gateway there is no issue, but directly from java agent it is giving that above issue.
- Firewall is opened for that port (as mentioned from collector agent i am able to send the traffic)
- I added debug logs and this is what it is showing -- 
this is the stack trace: 
[supportability_metrics_reporter] DEBUG io.opentelemetry.javaagent.shaded.instrumentation.api.internal.SupportabilityMetrics - Counter 'SqlStatementSanitizer cache miss' : 5
[OkHttp http://opera-tracing-gateway-np. trusted.com:4317/…I DEBUG okhttp3.internal. concurrent. TaskRunner - 010081 canceled
:OkHttp ConnectionPool
[OKHttp http://opera-tracing-gateway-np.trusted.com:4317/..) DEBUG OKhttp3. internal. concurrent. TaskRunner - 010013 scheduled after 0 us: OkHttp opera-tracing-gateway-np.trusted.com
[OkHttp http://opera-tracing-gateway-np. trusted.com:4317/..J DEBUG okhttp3. internal. concurrent. TaskRunner - 010001 scheduled after O us: OkHttp ConnectionPool
[OkHttp TaskRunner] DEBUG okhttp3.internal. concurrent. TaskRunner - 010013 starting
: OkHttp opera-tracing-gateway-np.trusted.com
[OkHttp TaskRunner] DEBUG okhttp3. internal. concurrent. TaskRunner - 010001 starting
:OkHttp ConnectionPool
[OkHttp ConnectionPool] DEBUG okhttp3.internal. concurrent. TaskRunner - 010001 run again after 300 s : OKHttp ConnectionPool [OkHttp TaskRunner] DEBUG okhttp3. internal. concurrent. TaskRunner - 010001 finished run in 903 us: OkHtt ConnectionPool
[OkHttp TaskRunner] DEBUG okhttp3.internal. concurrent. TaskRunner - Q10013 finished run in 1 ms: OkHttp opera-tracing-gateway-np.trusted.com
[OkHttp http://opera-tracing-gateway-np.trusted.com: 4317/..J DEBUG oKhttp3.internal. concurrent.TaskRunner - 010001 scheduled after O us: OkHttp ConnectionPool
[okHttp TaskRunner] DEBUG okhttp3. internal. concurrent. TaskRunner - Q10001 starting: OkHttp ConnectionPool
LOKHttp ConnectionPooll DEBUG oKhttp3. internal. concurrent. TaskRünner - 010001 run again after 300 s : OKHttp ConnectionPool
[OkHttp http://opera-tracing-gateway-np.trusted.com:4317/..J ERROR lo.opentelemetry.exporter.internal.grpc.OkHttpGrpcExporter-Failed to export spans.
The request could not be executed. Full error message: Required SETTINGS preface not received
[OkHttp TaskRunner] DEBUG okhttp3.internal. concurrent. TaskRunner - Q10001 finished run in 166 Ms: OKHttp ConnectionPool
[BatchSpanProcessor_WorkerThread-1] DEBUG lo.opentelemetry.sdk.trace.export. BatchSpanProcessor - Exporter failed

**Reply by trask** [reply]
can you confirm the JVM args you listed above?

> -Dotel. traces.exporter=otlp
-Dotel.exporter.otlp.traces.endpoint=:4317
-Dotel.exporter.otlp.traces.headers-"tIs.insecure=false"
-Dotel.service.name=petclinic

e.g. the endpoint looks incorrect, and the tls insecure also

**Reply by ramanjaneyagupta** [reply]
i am just playing with different diff options 
-Dotel.exporter.otlp.traces.endpoint=http://opera-tracing-gateway-np.trusted.com:4317

And as mentioned the same is working if i send to local collector and local collector to gateway - no issues but directly sending from the agent it is giving that above error, this is the same behaviour for both http/proto and grpc.

**Reply by ramanjaneyagupta** [reply]
and on collector side in logs it is saying in logs " received bogus greeting from the client"

**Reply by trask** [reply]
given that it's working with local collector, it seems like the Java agent is working correctly, and that it's some difference in the environment or setup of the remote collector

you could also try asking for advice in the collector repo

**Reply by ramanjaneyagupta** [reply]
@trask is there any way that we can send tIs.insecure, tls.insecure_skip_verify from the java agent ?

**Reply by trask** [reply]
`otel.exporter.otlp.traces.endpoint=https://host:port` will use tls

`otel.exporter.otlp.traces.endpoint=http://host:port` will not use tls

**Reply by ramanjaneyagupta** [reply]
Ok.. may be my question is a deviation.. even though my gateway cluster have tls enabled from my local collector on exporter option i am sending tls.insecure_skip_verify=true so that it is skipping the security while sending (or handshaking) now i am trying to send the traces directly to the gateway is there any option that we can send  "tIs.insecure=false, tls.insecure_skip_verify=true" these values from the agent? thanks!
